File: examples/python/subject_workflow.py
import json
import sys
from urllib.parse import quote

from memside_http import api_request


if len(sys.argv) != 3:
    raise RuntimeError(
        "Usage: python subject_workflow.py "
        "<non-secret-memory-id> <idempotency-key>"
    )

memory_id, idempotency_key = sys.argv[1:]
subject = api_request(
    "POST",
    "/subjects",
    body={
        "name": "Weekend reading",
        "subject_type": "topic",
        "aliases": ["Reading list"],
    },
)
subject_path = f"/subjects/{quote(subject['id'], safe='')}"
link = api_request(
    "POST",
    f"{subject_path}/memories",
    body={"memory_id": memory_id},
)
suggestion = api_request(
    "POST",
    f"{subject_path}/fact-suggestions",
    body={
        "suggestion_type": "add_fact",
        "proposed_fact_type": "custom",
        "proposed_fact_text": (
            "This item belongs to the weekend reading list."
        ),
        "source_memory_id": memory_id,
        "idempotency_key": idempotency_key,
    },
)

print(
    json.dumps(
        {"subject": subject, "link": link, "suggestion": suggestion},
        indent=2,
    )
)

File: examples/javascript/subject-workflow.mjs
import { memside } from "./memside-client.mjs";

const [memoryId, idempotencyKey] = process.argv.slice(2);
if (!memoryId || !idempotencyKey) {
  throw new Error(
    "Usage: node subject-workflow.mjs <non-secret-memory-id> <idempotency-key>"
  );
}

const subject = await memside("/subjects", {
  method: "POST",
  body: JSON.stringify({
    name: "Weekend reading",
    subject_type: "topic",
    aliases: ["Reading list"],
  }),
});

const encodedSubjectId = encodeURIComponent(subject.id);
const link = await memside(`/subjects/${encodedSubjectId}/memories`, {
  method: "POST",
  body: JSON.stringify({ memory_id: memoryId }),
});

const suggestion = await memside(
  `/subjects/${encodedSubjectId}/fact-suggestions`,
  {
    method: "POST",
    body: JSON.stringify({
      suggestion_type: "add_fact",
      proposed_fact_type: "custom",
      proposed_fact_text: "This item belongs to the weekend reading list.",
      source_memory_id: memoryId,
      idempotency_key: idempotencyKey,
    }),
  }
);

console.log(JSON.stringify({ subject, link, suggestion }, null, 2));
